sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -137,6 +143,21 @@ public ProcessBundleHandler(
         new UnknownPTransformRunnerFactory(urnToPTransformRunnerFactoryMap.keySet());
   }
 
+  private void extractMonitoringInfosToResponse(
```

Comment:
This method should build a list of MonitoringInfos and caller should set MonitoringInfos to response.
Doing this will simplify reading code greatly and prevent further complexity growth.

Reply:
Iterator would be good.

Reply:
Returning an iterable now, moved it to MetricContainerImpl